panic: close of closed channel, goroutine error.

I try to deploy python FastAPI app to Fly io and got goroutine error.

* MacOS
* fly v0.0.499 darwin/arm64 Commit: 10c95df6 BuildDate: 2023-03-24T04:50:43Z

** Paste your `fly.toml`

```toml
# fly.toml file generated for snowy-morning-6912 on 2023-03-25T19:27:29+07:00

app = "snowy-morning-6912"
kill_signal = "SIGINT"
kill_timeout = 5
primary_region = "sin"
processes = []

[build]
  builder = "paketobuildpacks/builder:base"

[env]
  PORT = "8080"

[experimental]
  auto_rollback = true

[[services]]
  http_checks = []
  internal_port = 8080
  processes = ["app"]
  protocol = "tcp"
  script_checks = []
  [services.concurrency]
    hard_limit = 25
    soft_limit = 20
    type = "connections"

  [[services.ports]]
    force_https = true
    handlers = ["http"]
    port = 80

  [[services.ports]]
    handlers = ["tls", "http"]
    port = 443

  [[services.tcp_checks]]
    grace_period = "1s"
    interval = "15s"
    restart_limit = 0
    timeout = "2s"

```

** Command output: **

```text

==> Verifying app config
--> Verified app config
==> Building image
Remote builder fly-builder-rough-wave-3915 ready
panic: close of closed channel

goroutine 93 [running]:
github.com/superfly/flyctl/internal/build/imgsrc.(*Resolver).StartHeartbeat.func2()
        github.com/superfly/flyctl/internal/build/imgsrc/resolver.go:507 +0x25c
created by github.com/superfly/flyctl/internal/build/imgsrc.(*Resolver).StartHeartbeat
        github.com/superfly/flyctl/internal/build/imgsrc/resolver.go:501 +0x548
```

--- a/internal/build/imgsrc/resolver.go
+++ b/internal/build/imgsrc/resolver.go
@@ -6,6 +6,7 @@ import (
 	"net"
 	"net/http"
 	"net/url"
+	"sync"
 	"time"
 
 	"github.com/pkg/errors"
@@ -61,6 +62,11 @@ type DeploymentImage struct {
 type Resolver struct {
 	dockerFactory *dockerClientFactory
 	apiClient     *api.Client
+}
+
+type StopSignal struct {
+	Chan chan struct{}
+	once sync.Once
 }
 
 // limit stored logs to 4KB; take suffix if longer
@@ -458,7 +464,7 @@ func (r *Resolver) finishBuild(ctx context.Context, build *build, failed bool, l
 
 // For remote builders send a periodic heartbeat during build to ensure machine stays alive
 // This is a noop for local builders
-func (r *Resolver) StartHeartbeat(ctx context.Context) chan<- interface{} {
+func (r *Resolver) StartHeartbeat(ctx context.Context) *StopSignal {
 	if !r.dockerFactory.remote {
 		return nil
 	}
@@ -495,15 +501,15 @@ func (r *Resolver) StartHeartbeat(ctx context.Context) chan<- interface{} {
 	pulseInterval := 30 * time.Second
 	maxTime := 1 * time.Hour
 
-	done := make(chan interface{})
-	time.AfterFunc(maxTime, func() { close(done) })
+	done := StopSignal{Chan: make(chan struct{})}
+	time.AfterFunc(maxTime, func() { done.Stop() })
 
 	go func() {
 		pulse := time.NewTicker(pulseInterval)
-		defer close(done)
+		defer done.Stop()
 		for {
 			select {
-			case <-done:
+			case <-done.Chan:
 				return
 			case <-ctx.Done():
 				return
@@ -518,7 +524,7 @@ func (r *Resolver) StartHeartbeat(ctx context.Context) chan<- interface{} {
 			}
 		}
 	}()
-	return done
+	return &done
 }
 
 func getHeartbeatUrl(dockerClient *dockerclient.Client) (string, error) {
@@ -535,10 +541,13 @@ func getHeartbeatUrl(dockerClient *dockerclient.Client) (string, error) {
 	return parsed.String(), nil
 }
 
-func (r *Resolver) StopHeartbeat(heartbeat chan<- interface{}) {
-	if heartbeat != nil {
-		heartbeat <- struct{}{}
+func (s *StopSignal) Stop() {
+	if s == nil {
+		return
 	}
+	s.once.Do(func() {
+		close(s.Chan)
+	})
 }
 
 func NewResolver(daemonType DockerDaemonType, apiClient *api.Client, appName string, iostreams *iostreams.IOStreams) *Resolver {

--- a/internal/command/deploy/deploy.go
+++ b/internal/command/deploy/deploy.go
@@ -391,7 +391,7 @@ func determineImage(ctx context.Context, appConfig *appconfig.Config) (img *imgs
 
 	// finally, build the image
 	heartbeat := resolver.StartHeartbeat(ctx)
-	defer resolver.StopHeartbeat(heartbeat)
+	defer heartbeat.Stop()
 	if img, err = resolver.BuildImage(ctx, io, opts); err == nil && img == nil {
 		err = errors.New("no image specified")
 	}

